### PR TITLE
CLOUD-42308 added azure rm support

### PR DESCRIPTION
**Pull request by doktoric**
@martonsereg or @akanto or @keyki pls

alt tag

### CODE REVIEW

File: cloud-api/src/main/java/com/sequenceiq/cloudbreak/cloud/ResourceConnector.java [hunk 1]
```diff
@@ -17,12 +17,12 @@
...
     List<CloudResourceStatus> check(AuthenticatedContext authenticatedContext, List<CloudResource> resources);
 
-    List<CloudResourceStatus> terminate(AuthenticatedContext authenticatedContext, List<CloudResource> cloudResources);
+    List<CloudResourceStatus> terminate(AuthenticatedContext authenticatedContext, CloudStack stack, List<CloudResource> cloudResources);
 
     //e.g. security rules
     List<CloudResourceStatus> update(AuthenticatedContext authenticatedContext, CloudStack stack, List<CloudResource> resources);
 
-    List<CloudResourceStatus> upscale(AuthenticatedContext authenticatedContext, CloudStack stack, List<CloudResource> resources, int adjustment);
+    List<CloudResourceStatus> upscale(AuthenticatedContext authenticatedContext, CloudStack stack, List<CloudResource> resourcer);
```
Comment (martonsereg) [inline cloud-api/src/main/java/com/sequenceiq/cloudbreak/cloud/ResourceConnector.java]: there is a typo here: *resources*

File: cloud-arm/build.gradle [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+dependencies {
+
+    compile project(':cloud-api')
+    compile project(':cloud-common')
+    testCompile project(':cloud-reactor')
+
+    compile group: 'org.slf4j',                     name: 'slf4j-api',                      version: slf4jApiVersion
+    compile group: 'org.springframework',           name: 'spring-context-support',         version: '4.0.3.RELEASE'
+    compile group: 'org.springframework.boot',      name: 'spring-boot-starter',            version: springBootVersion
```
Comment (akanto) [inline cloud-arm/build.gradle]: Spring boot is not really needed here.

File: cloud-arm/build.gradle [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    compile group: 'io.projectreactor',             name: 'reactor-core',                   version: eventBusVersion
+    compile group: 'io.projectreactor',             name: 'reactor-bus',                    version: eventBusVersion
```
Comment (akanto) [inline cloud-arm/build.gradle]: - reactor dependency not needed
- freemarker is duplicated, would and also would be better to externalise the version number
- guava is used in another modul, it is better to introduce a common variable for it
- compile ("com.sequenceiq:azure-rest-client:${azureRestClientVersion}") { exclude group: 'log4j' } has a different syntac than others
- mocikto shall be included as test dependency
- junit shall be included as test dependency

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/view/ArmNetworkView.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    public ArmNetworkView(Network network) {
+        this.network = network;
+    }
+
+    public String getSubnetCIDR() {
+        return network.getSubnet().getCidr();
+    }
+
+    public String getPublicNetId() {
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/view/ArmNetworkView.java]: Copy paste error from openstack

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmClient.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        return createAccess(armCredential);
+    }
+
+    public AzureRMClient createAccess(ArmCredentialView armCredential) {
+        return new AzureRMClient(armCredential.getTenantId(), armCredential.getAccesKey(), armCredential.getSecretKey(), armCredential.getSubscriptionId());
+    }
+
+    public String getStorageName(CloudCredential credential, String region) {
+        ArmCredentialView armCredential = new ArmCredentialView(credential);
+        String result = armCredential.getSubscriptionId().replaceAll("\\s+", "").replaceAll("-", "") + region.toLowerCase()
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmClient.java]: replaceAll().replaceAll can be written as relpaceAll(regexp, )

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmClient.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        if (result.length() > END_INDEX) {
+            return result.substring(result.length() - END_INDEX, result.length());
+        }
+        return result;
+    }
+
+    public String getStorageGroup(CloudCredential credential, String region) {
+        ArmCredentialView armCredential = new ArmCredentialView(credential);
+        String result = armCredential.getSubscriptionId().replaceAll("\\s+", "").replaceAll("-", "") + region.toLowerCase()
+                .replaceAll("_", "")
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmClient.java]: replaceAll().replaceAll can be written as relpaceAll(regexp, )

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmClient.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import com.sequenceiq.cloud.azure.client.AzureRMClient;
+import com.sequenceiq.cloudbreak.cloud.arm.view.ArmCredentialView;
+import com.sequenceiq.cloudbreak.cloud.event.context.AuthenticatedContext;
+import com.sequenceiq.cloudbreak.cloud.event.context.CloudContext;
+import com.sequenceiq.cloudbreak.cloud.model.CloudCredential;
+
+@Component
+public class ArmClient {
+
+    private static final int END_INDEX = 24;
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmClient.java]: What is END_INDEX, please use a better name or a comment

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmInstanceConnector.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+    @Inject
+    private ArmClient armClient;
+
+    @Inject
+    private ArmMetadataCollector armMetadataCollector;
+
+    @Override
+    public String getConsoleOutput(AuthenticatedContext authenticatedContext, CloudInstance vm) {
+        return "-----END SSH HOST KEY FINGERPRINTS-----";
```
Comment (martonsereg) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmInstanceConnector.java]: Is it intentional? If yes at least a comment should be added why is this needed.

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+    public String buildParameters(CloudCredential credential, Network network, Image image) {
+        try {
+            return processTemplateIntoString(freemarkerConfiguration.getTemplate(armTemplateParametersPath, "UTF-8"),  new HashMap<>());
+        } catch (IOException | TemplateException e) {
+            throw new RuntimeException("Failed to process the Arm TemplateParameterBuilder", e);
+        }
+    }
+
+    @VisibleForTesting
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java]: @Visible for testing shall be avoided, and can be used with a Context config + Mocking and inside the SomeApplicationContext shall contain a Freemarker config 

+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringApplicationConfiguration(classes = SomeApplicationContext.class)
+public class LaunchStackComponentTest {

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            model.put("ssh_key", armCredentialView.getPublicKey());
+            model.put("region", CloudRegion.valueOf(cloudStack.getRegion()).value());
+            model.put("subnet1Prefix", cloudStack.getNetwork().getSubnet().getCidr());
+            model.put("addressPrefix", cloudStack.getNetwork().getParameter("subnetCIDR", String.class));
+            model.put("groups", cloudStack.getGroups());
+            model.put("ports", cloudStack.getSecurity().getRules().get(0).getPorts());
+            model.put("port_protocol", cloudStack.getSecurity().getRules().get(0).getProtocol());
+            model.put("corecustomData",
+                    new String(
+                            Base64.encodeBase64(
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java]: Duplicated code, can be put into separate private method

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                                    String.format("%s", cloudStack.getImage().getUserData(InstanceGroupType.CORE))
+                                            .getBytes())));
+            model.put("gatewaycustomData",
+                    new String(
+                            Base64.encodeBase64(
+                                    String.format("%s", cloudStack.getImage().getUserData(InstanceGroupType.GATEWAY))
+                                            .getBytes())));
+            String generatedTemplate = processTemplateIntoString(freemarkerConfiguration.getTemplate(armTemplatePath, "UTF-8"), model);
+            LOGGER.debug("Generated Heat template: {}", generatedTemplate);
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java]: Not a heat template

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+            return generatedTemplate;
+        } catch (IOException | TemplateException e) {
+            throw new RuntimeException("Failed to process the Arm TemplateBuilder", e);
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java]: Instead of throwing Runtime, can we just have some own Exception which extends Runtime

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }
+
+    public String buildParameters(CloudCredential credential, Network network, Image image) {
+        try {
+            return processTemplateIntoString(freemarkerConfiguration.getTemplate(armTemplateParametersPath, "UTF-8"),  new HashMap<>());
+        } catch (IOException | TemplateException e) {
+            throw new RuntimeException("Failed to process the Arm TemplateParameterBuilder", e);
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/ArmTemplateBuilder.java]: Same as above

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/status/ArmInstanceStatus.java [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    private ArmInstanceStatus(String status) {
+        this.status = status;
+    }
+
+    public String getStatus() {
+        return status;
+    }
+
+    public static InstanceStatus get(String status) {
+        if (status.equals(STOPPED.getStatus())) {
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/status/ArmInstanceStatus.java]: The switch-case in ArmStackStatus.java looks nicer, than a multiple if statement.

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/poller/context/ResourceGroupDeleteCheckerContext.java [hunk 15]
```diff
@@ -0,0 +1,17 @@
+package com.sequenceiq.cloudbreak.cloud.arm.poller.context;
+
+import com.sequenceiq.cloudbreak.cloud.arm.view.ArmCredentialView;
+
+public class ResourceGroupDeleteCheckerContext extends ArmStatusCheckerContext {
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/poller/context/ResourceGroupDeleteCheckerContext.java]: Do we need this? Looks like it is the same as ResourceGroupCheckerContext

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/poller/context/VirtualMachineDeleteCheckerContext.java [hunk 16]
```diff
@@ -0,0 +1,24 @@
+package com.sequenceiq.cloudbreak.cloud.arm.poller.context;
+
+import com.sequenceiq.cloudbreak.cloud.arm.view.ArmCredentialView;
+
+public class VirtualMachineDeleteCheckerContext extends ArmStatusCheckerContext {
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/poller/context/VirtualMachineDeleteCheckerContext.java]: Can we just use only one class instead of  VirtualMachineCheckerContext   + VirtualMachineDeleteCheckerContext

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/poller/context/NetworkInterfaceDeleteCheckerContext.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
+package com.sequenceiq.cloudbreak.cloud.arm.poller.context;
+
+import com.sequenceiq.cloudbreak.cloud.arm.view.ArmCredentialView;
+
+public class NetworkInterfaceDeleteCheckerContext extends ArmStatusCheckerContext {
+
+    private String groupName;
+    private String networkName;
+
+    public NetworkInterfaceDeleteCheckerContext(ArmCredentialView armCredentialView, String groupName, String networkName) {
```
Comment (akanto) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/poller/context/NetworkInterfaceDeleteCheckerContext.java]: Instead of NetworkInterfaceDeleteCheckerContex, maybe the NetworkInterfaceCheckerContext would be sufficient. And also in case of other Context , maybe the 'Delete' word could be eliminated

File: cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/poller/context/ResourceGroupDeleteCheckerContext.java (hunk 15 above)
Comment (doktoric) [inline cloud-arm/src/main/java/com/sequenceiq/cloudbreak/cloud/arm/poller/context/ResourceGroupDeleteCheckerContext.java]: deleted

File: cloud-common/src/main/java/com/sequenceiq/cloudbreak/poller/StatusCheckerTask.java [hunk 18]
```diff
@@ -1,4 +1,4 @@
-package com.sequenceiq.cloudbreak.service;
+package com.sequenceiq.cloudbreak.poller;
 
 public interface StatusCheckerTask<T> {
```
Comment (akanto) [inline cloud-common/src/main/java/com/sequenceiq/cloudbreak/poller/StatusCheckerTask.java]: There is a PollTask.java in the cloud-api, please use that here, instead of introducing StatusCheckerTask
